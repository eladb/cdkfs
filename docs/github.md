# GitHub

By default, many projects are initialized with `GitHub` component to enabled GitHub as the default provider for CI/CD workflows. See https://docs.github.com/en/actions for more information.

The use of GitHub (and generating corresponding files in `.github`) can be disabled by specifying `github: false` in your project options.

## Workflows

TODO

See the `GitHub`, `GithubWorkflow`, and `Job` types in the [API reference](./api/API.md) for currently available APIs.

Example code of creating a GitHub workflow: <https://github.com/projen/projen/blob/65b4194c163f47ba4842981b0c92dbe516be787b/src/github/auto-approve.ts#L67-L105>

### Stale workflow

A "stale" workflow can be added which will automatically close issues or pull
requests on your GitHub repository after time has passed without the issue
seeing any comments or updates. You can enable as shown below:

```ts
// or PythonProject, etc.
new typescript.TypeScriptProject({
  stale: true,
  staleOptions: {
    issues: {
      closeMessage: "closing pull request",
      staleLabel: "I-AM-STALE",
      daysBeforeStale: 180,
    },
  }
})
```

Check the API reference for a list of all available options.

When enabled, by default issues with no activity with will be marked as stale
after 60 days and closed within 7 days, and pull requests with no activity will
be marked as stale after 14 days and closed within 2 days.
