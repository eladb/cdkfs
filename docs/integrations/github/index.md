---
sidebar_position: 1
---

# GitHub Integration


This is just a placeholder doc. You can contribute to this project! Please see the [Contributing Guide](https://github.com/projen/projen/blob/main/docusaurus/README.md) for more information.
