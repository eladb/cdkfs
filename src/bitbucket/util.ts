

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}