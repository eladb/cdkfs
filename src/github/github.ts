import { Component } from '../component';
import { FileBase } from '../file';
import { Project } from '../project';
import { Dependabot, DependabotOptions } from './dependabot';
import { Mergify } from './mergify';
import { PullRequestTemplate } from './pr-template';
import { GithubWorkflow } from './workflows';

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;
}

export class GitHub extends Component {
  /**
   * The `Mergify` configured on this repository. This is `undefined` if Mergify
   * was not enabled when creating the repository.
   */
  public readonly mergify?: Mergify;

  /**
   * The .gitattributes file for this repository.
   */
  private readonly gitattributes: GitattributesFile;

  public constructor(project: Project, options: GitHubOptions) {
    super(project);

    this.gitattributes = new GitattributesFile(project);

    // Contents  of the .projen/ directory are generated by projen
    this.annotateGenerated('/.projen/**');
    // The .gitattributes file itself is generated
    this.annotateGenerated(`/${this.gitattributes.path}`);

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this);
    }
  }

  public addWorkflow(name: string) {
    return new GithubWorkflow(this, name);
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }

  /**
   * Marks the provided file(s) as being generated. This is achieved using the
   * github-linguist attributes. Generated files do not count against the
   * repository statistics and language breakdown.
   *
   * @param glob the glob pattern to match (could be a file path).
   *
   * @see https://github.com/github/linguist/blob/master/docs/overrides.md
   */
  public annotateGenerated(glob: string): void {
    this.gitattributes.addAttributes(glob, 'linguist-generated');
  }
}

class GitattributesFile extends FileBase {
  private readonly attributes = new Map<string, Set<string>>();

  public constructor(project: Project) {
    super(project, '.gitattributes');
  }

  public addAttributes(glob: string, ...attributes: readonly string[]) {
    if (!this.attributes.has(glob)) {
      this.attributes.set(glob, new Set());
    }
    const set = this.attributes.get(glob)!;
    for (const attribute of attributes) {
      set.add(attribute);
    }
  }

  protected synthesizeContent(): string | undefined {
    // We can assume the file map is never empty.
    const entries = Array.from(this.attributes.entries())
      .sort(([l], [r]) => l.localeCompare(r));

    const maxLen = Math.max(...entries.map(([glob]) => glob.length));
    return [
      `# ${FileBase.PROJEN_MARKER}`,
      '',
      ...entries.map(([name, attributes]) => `${name.padEnd(maxLen, ' ')}\t${Array.from(attributes).join(' ')}`),
    ].join('\n');
  }
}
