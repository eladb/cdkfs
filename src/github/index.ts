export * from './github';

export * from './dependabot';
export * from './workflows';
export * from './mergify';
export * from './pr-template';
export * from './mergify';
export * from './auto-merge';
export * from './auto-approve';
export * from './stale';

export * from './task-workflow';

export * as workflows from './workflows-model';
