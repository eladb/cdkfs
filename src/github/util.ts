export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
