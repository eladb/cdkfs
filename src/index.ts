export * from './file';
export * from './github-workflow';
export * from './ignore-file';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './node-project';
export * from './project';
export * from './component';
export * from './semver';
export * from './version';
export * from './eslint';
export * from './jest';
export * from './typescript';
export * from './mergify';
export * from './dependabot';