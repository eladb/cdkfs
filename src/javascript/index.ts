export * from './bundler';
export * from './eslint';
export * from './jest';
export * from './node-package';
export * from './node-project';
export * from './npm-config';
export * from './upgrade-dependencies';
export * from './projenrc';
export * from './typescript-config';
