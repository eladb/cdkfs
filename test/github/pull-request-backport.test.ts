// import * as yaml from "yaml";
import { PullRequestBackport } from "../../src/github";
import { NodeProject, NodeProjectOptions } from "../../src/javascript";
import { synthSnapshot } from "../util";

test("default", () => {
  // GIVEN
  const project = createProject();

  // WHEN
  new PullRequestBackport(project);

  // THEN
  const snapshot = synthSnapshot(project);

  const workflow = snapshot[".github/workflows/backport.yml"];
  expect(workflow).toBeDefined();
  expect(workflow).toMatchSnapshot("workflow");
  expect(workflow).toContain("Backport PR");
  expect(workflow).toContain("auto_backport_label_prefix: backport-to-");

  const settings = snapshot[".backportrc.json"];
  expect(settings).toMatchSnapshot("settings");
  expect(settings.backportBranchName).toMatch("backport/");
  expect(settings.commitConflicts).toBe(true);
  expect(settings.targetPRLabels).toEqual(["backport", "auto-approve"]);
  expect(settings.targetBranchChoices).toEqual(["main"]);
});

test("non-default settings", () => {
  // GIVEN
  const project = createProject();

  // WHEN
  new PullRequestBackport(project, {
    workflowName: "other-backport",
    createWithConflicts: false,
    backportPRLabels: ["magic"],
    backportBranchNamePrefix: "auto-backport/",
    autoApproveBackport: false,
    branches: ["one", "two"],
    labelPrefix: "copy-to-",
  });

  // THEN
  const snapshot = synthSnapshot(project);

  const workflow = snapshot[".github/workflows/other-backport.yml"];
  expect(workflow).toBeDefined();
  expect(workflow).toMatchSnapshot("workflow");
  expect(workflow).toContain("auto_backport_label_prefix: copy-to-");

  const settings = snapshot[".backportrc.json"];
  expect(settings).toMatchSnapshot("settings");
  expect(settings.commitConflicts).toBe(false);
  expect(settings.backportBranchName).toMatch("auto-backport/");
  expect(settings.targetPRLabels).toEqual(["magic"]);
  expect(settings.targetBranchChoices).toEqual(["one", "two"]);
});

test("all release branches are automatically added", () => {
  // GIVEN
  const project = createProject({
    majorVersion: 3,
    releaseBranches: {
      v1: {
        majorVersion: 1,
      },
      v2: {
        majorVersion: 2,
      },
    },
  });

  // WHEN
  new PullRequestBackport(project);

  // THEN
  const snapshot = synthSnapshot(project);
  const settings = snapshot[".backportrc.json"];
  expect(settings.targetBranchChoices).toEqual(["main", "v1", "v2"]);
});

describe("no labels on backport PRs", () => {
  it("uses the backport branch name to detect backport PRs", () => {
    // GIVEN
    const project = createProject();

    // WHEN
    new PullRequestBackport(project, {
      backportPRLabels: [],
    });

    // THEN
    const snapshot = synthSnapshot(project);
    const workflow = snapshot[".github/workflows/backport.yml"];
    expect(workflow).toBeDefined();
    expect(workflow).toMatchSnapshot("workflow");
    expect(workflow).toContain("!startsWith(github.head_ref, 'backport/')");
  });
});

test("can be added via githubOptions", () => {
  // GIVEN
  const project = createProject({
    githubOptions: {
      pullRequestBackport: true,
      pullRequestBackportOptions: {
        backportPRLabels: ["magic"],
      },
    },
  });
  // THEN
  const snapshot = synthSnapshot(project);

  const workflow = snapshot[".github/workflows/backport.yml"];
  expect(workflow).toBeDefined();
  expect(workflow).toContain("Backport PR");

  const settings = snapshot[".backportrc.json"];
  expect(settings.targetPRLabels).toEqual(["magic"]);
});

type ProjectOptions = Omit<
  NodeProjectOptions,
  "outdir" | "defaultReleaseBranch" | "name"
>;
function createProject(options: ProjectOptions = {}): NodeProject {
  return new NodeProject({
    defaultReleaseBranch: "main",
    name: "node-project",
    githubOptions: {
      pullRequestLintOptions: {
        semanticTitle: false,
      },
    },
    autoApproveOptions: {},
    ...options,
  });
}
