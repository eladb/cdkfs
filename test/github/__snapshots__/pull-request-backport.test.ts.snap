// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default: settings 1`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "backportBranchName": "backport/{{targetBranch}}-{{refValues}}",
  "commitConflicts": true,
  "prTitle": "{{sourcePullRequest.title}} (backport #{{sourcePullRequest.number}})",
  "targetBranchChoices": [
    "main",
  ],
  "targetPRLabels": [
    "backport",
    "auto-approve",
  ],
}
`;

exports[`default: workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: backport
on:
  pull_request_target:
    types:
      - labeled
      - closed
jobs:
  backport:
    name: Backport PR
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - name: Backport Action
        if: github.event.pull_request.merged == true && !(contains(github.event.pull_request.labels.*.name, 'backport'))
        uses: sqren/backport-github-action@v9.5.1
        with:
          github_token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          auto_backport_label_prefix: backport-to-
"
`;

exports[`no labels on backport PRs uses the backport branch name to detect backport PRs: workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: backport
on:
  pull_request_target:
    types:
      - labeled
      - closed
jobs:
  backport:
    name: Backport PR
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - name: Backport Action
        if: github.event.pull_request.merged == true && !startsWith(github.head_ref, 'backport/'))
        uses: sqren/backport-github-action@v9.5.1
        with:
          github_token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          auto_backport_label_prefix: backport-to-
"
`;

exports[`non-default settings: settings 1`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "backportBranchName": "auto-backport/{{targetBranch}}-{{refValues}}",
  "commitConflicts": false,
  "prTitle": "{{sourcePullRequest.title}} (backport #{{sourcePullRequest.number}})",
  "targetBranchChoices": [
    "one",
    "two",
  ],
  "targetPRLabels": [
    "magic",
  ],
}
`;

exports[`non-default settings: workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: other-backport
on:
  pull_request_target:
    types:
      - labeled
      - closed
jobs:
  backport:
    name: Backport PR
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - name: Backport Action
        if: github.event.pull_request.merged == true && !(contains(github.event.pull_request.labels.*.name, 'magic'))
        uses: sqren/backport-github-action@v9.5.1
        with:
          github_token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          auto_backport_label_prefix: copy-to-
"
`;
